### DISCUSSION TITLE
sqlalchemy.exc.OperationalError: (pymysql.err.OperationalError)

### DISCUSSION THREAD

**Discussion by tariqjamal-abdulazeez** [miguelgrinberg/Flask-Migrate#532] [Q&A] [answered]
`flask db upgrade`

> 'FLASK_ENV' is deprecated and will not be used in Flask 2.3. Use 'FLASK_DEBUG' instead.
> 'FLASK_ENV' is deprecated and will not be used in Flask 2.3. Use 'FLASK_DEBUG' instead.
> INFO  [alembic.env] Tables which are allowed for autogeneration: vsessions, vparticipants, vrecordings, vevents, interpreter_vsession_notifications, provider_vri_settings, notification_device, sites, customers, interpreters, languages, contacts, customers, logins_int, interpreter_availability, mobile_app_versions, operators
> INFO  [alembic.env] To add more, look at alembic.ini in the alembic:include section
> Traceback (most recent call last):
>   File "/Users/tariqjamal/anaconda3/envs/lvdispatch_2/lib/python3.9/site-packages/sqlalchemy/engine/base.py", line 145, in __init__
>     self._dbapi_connection = engine.raw_connection()
>   File "/Users/tariqjamal/anaconda3/envs/lvdispatch_2/lib/python3.9/site-packages/sqlalchemy/engine/base.py", line 3292, in raw_connection
>     return self.pool.connect()
>   File "/Users/tariqjamal/anaconda3/envs/lvdispatch_2/lib/python3.9/site-packages/sqlalchemy/pool/base.py", line 452, in connect
>     return _ConnectionFairy._checkout(self)
>   File "/Users/tariqjamal/anaconda3/envs/lvdispatch_2/lib/python3.9/site-packages/sqlalchemy/pool/base.py", line 1269, in _checkout
>     fairy = _ConnectionRecord.checkout(pool)
>   File "/Users/tariqjamal/anaconda3/envs/lvdispatch_2/lib/python3.9/site-packages/sqlalchemy/pool/base.py", line 716, in checkout
>     rec = pool._do_get()
>   File "/Users/tariqjamal/anaconda3/envs/lvdispatch_2/lib/python3.9/site-packages/sqlalchemy/pool/impl.py", line 284, in _do_get
>     return self._create_connection()
>   File "/Users/tariqjamal/anaconda3/envs/lvdispatch_2/lib/python3.9/site-packages/sqlalchemy/pool/base.py", line 393, in _create_connection
>     return _ConnectionRecord(self)
>   File "/Users/tariqjamal/anaconda3/envs/lvdispatch_2/lib/python3.9/site-packages/sqlalchemy/pool/base.py", line 678, in __init__
>     self.__connect()
>   File "/Users/tariqjamal/anaconda3/envs/lvdispatch_2/lib/python3.9/site-packages/sqlalchemy/pool/base.py", line 903, in __connect
>     pool.logger.debug("Error on connect(): %s", e)
>   File "/Users/tariqjamal/anaconda3/envs/lvdispatch_2/lib/python3.9/site-packages/sqlalchemy/util/langhelpers.py", line 146, in __exit__
>     raise exc_value.with_traceback(exc_tb)
>   File "/Users/tariqjamal/anaconda3/envs/lvdispatch_2/lib/python3.9/site-packages/sqlalchemy/pool/base.py", line 898, in __connect
>     self.dbapi_connection = connection = pool._invoke_creator(self)
>   File "/Users/tariqjamal/anaconda3/envs/lvdispatch_2/lib/python3.9/site-packages/sqlalchemy/engine/create.py", line 637, in connect
>     return dialect.connect(*cargs, **cparams)
>   File "/Users/tariqjamal/anaconda3/envs/lvdispatch_2/lib/python3.9/site-packages/sqlalchemy/engine/default.py", line 616, in connect
>     return self.loaded_dbapi.connect(*cargs, **cparams)
>   File "/Users/tariqjamal/anaconda3/envs/lvdispatch_2/lib/python3.9/site-packages/pymysql/__init__.py", line 94, in Connect
>     return Connection(*args, **kwargs)
>   File "/Users/tariqjamal/anaconda3/envs/lvdispatch_2/lib/python3.9/site-packages/pymysql/connections.py", line 325, in __init__
>     self.connect()
>   File "/Users/tariqjamal/anaconda3/envs/lvdispatch_2/lib/python3.9/site-packages/pymysql/connections.py", line 599, in connect
>     self._request_authentication()
>   File "/Users/tariqjamal/anaconda3/envs/lvdispatch_2/lib/python3.9/site-packages/pymysql/connections.py", line 871, in _request_authentication
>     auth_packet = self._process_auth(plugin_name, auth_packet)
>   File "/Users/tariqjamal/anaconda3/envs/lvdispatch_2/lib/python3.9/site-packages/pymysql/connections.py", line 942, in _process_auth
>     pkt = self._read_packet()
>   File "/Users/tariqjamal/anaconda3/envs/lvdispatch_2/lib/python3.9/site-packages/pymysql/connections.py", line 684, in _read_packet
>     packet.check_error()
>   File "/Users/tariqjamal/anaconda3/envs/lvdispatch_2/lib/python3.9/site-packages/pymysql/protocol.py", line 220, in check_error
>     err.raise_mysql_exception(self._data)
>   File "/Users/tariqjamal/anaconda3/envs/lvdispatch_2/lib/python3.9/site-packages/pymysql/err.py", line 109, in raise_mysql_exception
>     raise errorclass(errno, errval)
> pymysql.err.OperationalError: (1045, "Access denied for user 'root'@'localhost' (using password: YES)")
> 
> The above exception was the direct cause of the following exception:
> 
> Traceback (most recent call last):
>   File "/Users/tariqjamal/anaconda3/envs/lvdispatch_2/bin/flask", line 8, in <module>
>     sys.exit(main())
>   File "/Users/tariqjamal/anaconda3/envs/lvdispatch_2/lib/python3.9/site-packages/flask/cli.py", line 1050, in main
>     cli.main()
>   File "/Users/tariqjamal/anaconda3/envs/lvdispatch_2/lib/python3.9/site-packages/click/core.py", line 1078, in main
>     rv = self.invoke(ctx)
>   File "/Users/tariqjamal/anaconda3/envs/lvdispatch_2/lib/python3.9/site-packages/click/core.py", line 1688, in invoke
>     return _process_result(sub_ctx.command.invoke(sub_ctx))
>   File "/Users/tariqjamal/anaconda3/envs/lvdispatch_2/lib/python3.9/site-packages/click/core.py", line 1688, in invoke
>     return _process_result(sub_ctx.command.invoke(sub_ctx))
>   File "/Users/tariqjamal/anaconda3/envs/lvdispatch_2/lib/python3.9/site-packages/click/core.py", line 1434, in invoke
>     return ctx.invoke(self.callback, **ctx.params)
>   File "/Users/tariqjamal/anaconda3/envs/lvdispatch_2/lib/python3.9/site-packages/click/core.py", line 783, in invoke
>     return __callback(*args, **kwargs)
>   File "/Users/tariqjamal/anaconda3/envs/lvdispatch_2/lib/python3.9/site-packages/click/decorators.py", line 33, in new_func
>     return f(get_current_context(), *args, **kwargs)
>   File "/Users/tariqjamal/anaconda3/envs/lvdispatch_2/lib/python3.9/site-packages/flask/cli.py", line 357, in decorator
>     return __ctx.invoke(f, *args, **kwargs)
>   File "/Users/tariqjamal/anaconda3/envs/lvdispatch_2/lib/python3.9/site-packages/click/core.py", line 783, in invoke
>     return __callback(*args, **kwargs)
>   File "/Users/tariqjamal/anaconda3/envs/lvdispatch_2/lib/python3.9/site-packages/flask_migrate/cli.py", line 150, in upgrade
>     _upgrade(directory, revision, sql, tag, x_arg)
>   File "/Users/tariqjamal/anaconda3/envs/lvdispatch_2/lib/python3.9/site-packages/flask_migrate/__init__.py", line 111, in wrapped
>     f(*args, **kwargs)
>   File "/Users/tariqjamal/anaconda3/envs/lvdispatch_2/lib/python3.9/site-packages/flask_migrate/__init__.py", line 200, in upgrade
>     command.upgrade(config, revision, sql=sql, tag=tag)
>   File "/Users/tariqjamal/anaconda3/envs/lvdispatch_2/lib/python3.9/site-packages/alembic/command.py", line 398, in upgrade
>     script.run_env()
>   File "/Users/tariqjamal/anaconda3/envs/lvdispatch_2/lib/python3.9/site-packages/alembic/script/base.py", line 579, in run_env
>     util.load_python_file(self.dir, "env.py")
>   File "/Users/tariqjamal/anaconda3/envs/lvdispatch_2/lib/python3.9/site-packages/alembic/util/pyfiles.py", line 93, in load_python_file
>     module = load_module_py(module_id, path)
>   File "/Users/tariqjamal/anaconda3/envs/lvdispatch_2/lib/python3.9/site-packages/alembic/util/pyfiles.py", line 109, in load_module_py
>     spec.loader.exec_module(module)  # type: ignore
>   File "<frozen importlib._bootstrap_external>", line 850, in exec_module
>   File "<frozen importlib._bootstrap>", line 228, in _call_with_frames_removed
>   File "migrations/env.py", line 123, in <module>
>     run_migrations_online()
>   File "migrations/env.py", line 105, in run_migrations_online
>     with connectable.connect() as connection:
>   File "/Users/tariqjamal/anaconda3/envs/lvdispatch_2/lib/python3.9/site-packages/sqlalchemy/engine/base.py", line 3268, in connect
>     return self._connection_cls(self)
>   File "/Users/tariqjamal/anaconda3/envs/lvdispatch_2/lib/python3.9/site-packages/sqlalchemy/engine/base.py", line 147, in __init__
>     Connection._handle_dbapi_exception_noconnection(
>   File "/Users/tariqjamal/anaconda3/envs/lvdispatch_2/lib/python3.9/site-packages/sqlalchemy/engine/base.py", line 2430, in _handle_dbapi_exception_noconnection
>     raise sqlalchemy_exception.with_traceback(exc_info[2]) from e
>   File "/Users/tariqjamal/anaconda3/envs/lvdispatch_2/lib/python3.9/site-packages/sqlalchemy/engine/base.py", line 145, in __init__
>     self._dbapi_connection = engine.raw_connection()
>   File "/Users/tariqjamal/anaconda3/envs/lvdispatch_2/lib/python3.9/site-packages/sqlalchemy/engine/base.py", line 3292, in raw_connection
>     return self.pool.connect()
>   File "/Users/tariqjamal/anaconda3/envs/lvdispatch_2/lib/python3.9/site-packages/sqlalchemy/pool/base.py", line 452, in connect
>     return _ConnectionFairy._checkout(self)
>   File "/Users/tariqjamal/anaconda3/envs/lvdispatch_2/lib/python3.9/site-packages/sqlalchemy/pool/base.py", line 1269, in _checkout
>     fairy = _ConnectionRecord.checkout(pool)
>   File "/Users/tariqjamal/anaconda3/envs/lvdispatch_2/lib/python3.9/site-packages/sqlalchemy/pool/base.py", line 716, in checkout
>     rec = pool._do_get()
>   File "/Users/tariqjamal/anaconda3/envs/lvdispatch_2/lib/python3.9/site-packages/sqlalchemy/pool/impl.py", line 284, in _do_get
>     return self._create_connection()
>   File "/Users/tariqjamal/anaconda3/envs/lvdispatch_2/lib/python3.9/site-packages/sqlalchemy/pool/base.py", line 393, in _create_connection
>     return _ConnectionRecord(self)
>   File "/Users/tariqjamal/anaconda3/envs/lvdispatch_2/lib/python3.9/site-packages/sqlalchemy/pool/base.py", line 678, in __init__
>     self.__connect()
>   File "/Users/tariqjamal/anaconda3/envs/lvdispatch_2/lib/python3.9/site-packages/sqlalchemy/pool/base.py", line 903, in __connect
>     pool.logger.debug("Error on connect(): %s", e)
>   File "/Users/tariqjamal/anaconda3/envs/lvdispatch_2/lib/python3.9/site-packages/sqlalchemy/util/langhelpers.py", line 146, in __exit__
>     raise exc_value.with_traceback(exc_tb)
>   File "/Users/tariqjamal/anaconda3/envs/lvdispatch_2/lib/python3.9/site-packages/sqlalchemy/pool/base.py", line 898, in __connect
>     self.dbapi_connection = connection = pool._invoke_creator(self)
>   File "/Users/tariqjamal/anaconda3/envs/lvdispatch_2/lib/python3.9/site-packages/sqlalchemy/engine/create.py", line 637, in connect
>     return dialect.connect(*cargs, **cparams)
>   File "/Users/tariqjamal/anaconda3/envs/lvdispatch_2/lib/python3.9/site-packages/sqlalchemy/engine/default.py", line 616, in connect
>     return self.loaded_dbapi.connect(*cargs, **cparams)
>   File "/Users/tariqjamal/anaconda3/envs/lvdispatch_2/lib/python3.9/site-packages/pymysql/__init__.py", line 94, in Connect
>     return Connection(*args, **kwargs)
>   File "/Users/tariqjamal/anaconda3/envs/lvdispatch_2/lib/python3.9/site-packages/pymysql/connections.py", line 325, in __init__
>     self.connect()
>   File "/Users/tariqjamal/anaconda3/envs/lvdispatch_2/lib/python3.9/site-packages/pymysql/connections.py", line 599, in connect
>     self._request_authentication()
>   File "/Users/tariqjamal/anaconda3/envs/lvdispatch_2/lib/python3.9/site-packages/pymysql/connections.py", line 871, in _request_authentication
>     auth_packet = self._process_auth(plugin_name, auth_packet)
>   File "/Users/tariqjamal/anaconda3/envs/lvdispatch_2/lib/python3.9/site-packages/pymysql/connections.py", line 942, in _process_auth
>     pkt = self._read_packet()
>   File "/Users/tariqjamal/anaconda3/envs/lvdispatch_2/lib/python3.9/site-packages/pymysql/connections.py", line 684, in _read_packet
>     packet.check_error()
>   File "/Users/tariqjamal/anaconda3/envs/lvdispatch_2/lib/python3.9/site-packages/pymysql/protocol.py", line 220, in check_error
>     err.raise_mysql_exception(self._data)
>   File "/Users/tariqjamal/anaconda3/envs/lvdispatch_2/lib/python3.9/site-packages/pymysql/err.py", line 109, in raise_mysql_exception
>     raise errorclass(errno, errval)
> sqlalchemy.exc.OperationalError: (pymysql.err.OperationalError) (1045, "Access denied for user 'root'@'localhost' (using password: YES)")
> (Background on this error at: https://sqlalche.me/e/20/e3q8)`

This error appears when I try to migrate changes to the database; however, when I make any API call, the data is rendered correctly. The above error only appears when I try to migrate changes to the database using `flask db migrate`.

 **Packages:**
`poetry show flask-migrate`
> name         : flask-migrate                                                       
> version      : 2.7.0                                                               
> description  : SQLAlchemy database migrations for Flask applications using Alembic 
> 
> dependencies
>  - alembic >=0.7
>  - Flask >=0.9
>  - Flask-SQLAlchemy >=1.0

`poetry show flask-sqlalchemy`
> name         : flask-sqlalchemy                                  
> version      : 3.1.1                                             
> description  : Add SQLAlchemy support to your Flask application. 
> 
> dependencies
>  - flask >=2.2.5
>  - sqlalchemy >=2.0.16
> 
> required by
>  - flask-migrate >=1.0

**Comment by miguelgrinberg** [comment]
This is a database error, nothing to do with Flask-Migrate:

> Access denied for user 'root'@'localhost' (using password: YES)

Make sure that your credentials and/or database URL are correctly configured.

**Reply by tariqjamal-abdulazeez** [reply]
@miguelgrinberg  I had correctly configured the db connection with Api and if I make any api it will perfectly rendering the data from database and  this was occurring after I upgrade the flask-sqlalchemy version from 2.5.3 to 3.1.1 in flask-sqlalchemy==2.5.3 when i run `flask db migrate` it  will perfectly working without any but after upgrading flask-sqlalchemy==3.1.1 getting the error when I run migrate

**Reply by miguelgrinberg** [reply, marked as answer]
Try updating the env.py file that is in your `migrations` directory. You can find the current version here: https://github.com/miguelgrinberg/Flask-Migrate/tree/main/src/flask_migrate/templates/flask

**Reply by tariqjamal-abdulazeez** [reply]
@miguelgrinberg  Can you specify what we need to change in this file? [https://github.com/miguelgrinberg/Flask-Migrate/blob/main/src/flask_migrate/templates/flask/env.py](url) in order to make it work, Because I attempted to make some changes in this, but they did not work for me.

**Reply by miguelgrinberg** [reply]
If you do not have any custom changes in this file, you can just replace it with the version that I linked.

**Reply by tariqjamal-abdulazeez** [reply]
I went over the migration logic again step by step. After that, I made changes to the env.py file, and it worked for me.
